File: rag/web_search.py
"""
DuckDuckGo web search wrapper with personality-filtered queries.
Falls back gracefully if search fails.
"""
import logging

from duckduckgo_search import DDGS

logger = logging.getLogger(__name__)

# Personality-biased search prefixes
_PERSONALITY_PREFIXES: dict[str, str] = {
    "elon_musk": "engineering physics first-principles",
    "sam_altman": "AI startups exponential-growth",
    "dario_amodei": "AI-safety research empirical",
    "mark_zuckerberg": "platform open-source product-strategy",
}


def build_search_query(speaker: str, question: str) -> str:
    """Prepend personality perspective terms to bias search results."""
    prefix = _PERSONALITY_PREFIXES.get(speaker, "")
    return f"{prefix} {question}".strip()


def search_web(query: str, speaker: str, max_results: int = 5) -> list[dict]:
    """Run a DuckDuckGo search. Returns list of {title, href, body} dicts.

    Wrapped in try/except so failures degrade gracefully to Pinecone-only.
    """
    try:
        with DDGS() as ddgs:
            results = list(ddgs.text(query, max_results=max_results))
        return results
    except Exception as exc:
        logger.warning("Web search failed for query %r: %s", query, exc)
        return []


def format_web_results(results: list[dict], max_words: int = 500) -> str:
    """Format search results into prompt-ready text, capped at ~max_words."""
    if not results:
        return ""

    lines: list[str] = []
    word_count = 0
    for r in results:
        title = r.get("title", "")
        body = r.get("body", "")
        href = r.get("href", "")
        entry = f"- {title}: {body} (source: {href})"
        entry_words = len(entry.split())
        if word_count + entry_words > max_words:
            break
        lines.append(entry)
        word_count += entry_words

    return "\n".join(lines)

File: ceo-arena-next/lib/rag/web-search.ts
const PERSONALITY_PREFIXES: Record<string, string> = {
  elon_musk: "engineering physics first-principles",
  sam_altman: "AI startups exponential-growth",
  dario_amodei: "AI-safety research empirical",
  mark_zuckerberg: "platform open-source product-strategy",
};

export function buildSearchQuery(speaker: string, question: string): string {
  const prefix = PERSONALITY_PREFIXES[speaker] ?? "";
  return `${prefix} ${question}`.trim();
}

interface SearchResult {
  title: string;
  href: string;
  body: string;
}

export async function searchWeb(
  query: string,
  maxResults = 5,
): Promise<SearchResult[]> {
  try {
    const url = `https://html.duckduckgo.com/html/?q=${encodeURIComponent(query)}`;
    const response = await fetch(url, {
      headers: {
        "User-Agent": "Mozilla/5.0 (compatible; CEOArena/1.0)",
      },
      signal: AbortSignal.timeout(10000),
    });

    if (!response.ok) return [];

    const html = await response.text();
    const results: SearchResult[] = [];

    const resultRegex = /<a[^>]+class="result__a"[^>]+href="([^"]*)"[^>]*>([\s\S]*?)<\/a>[\s\S]*?<a[^>]+class="result__snippet"[^>]*>([\s\S]*?)<\/a>/g;
    let match: RegExpExecArray | null;
    while ((match = resultRegex.exec(html)) !== null && results.length < maxResults) {
      const href = decodeURIComponent((match[1] ?? "").replace(/.*uddg=/, "").replace(/&.*/, ""));
      const title = (match[2] ?? "").replace(/<[^>]*>/g, "").trim();
      const body = (match[3] ?? "").replace(/<[^>]*>/g, "").trim();
      if (title && body) {
        results.push({ title, href, body });
      }
    }

    return results;
  } catch {
    return [];
  }
}

export function formatWebResults(results: SearchResult[], maxWords = 500): string {
  if (results.length === 0) return "";

  const lines: string[] = [];
  let wordCount = 0;

  for (const r of results) {
    const entry = `- ${r.title}: ${r.body} (source: ${r.href})`;
    const entryWords = entry.split(/\s+/).length;
    if (wordCount + entryWords > maxWords) break;
    lines.push(entry);
    wordCount += entryWords;
  }

  return lines.join("\n");
}
